### DISCUSSION TITLE
System "health" monitoring?

### DISCUSSION THREAD

**Discussion by gmcmicken** [Arnold-n/P1P2MQTT#44] [Ideas]
Longer term, I am thinking about using P1P2 library to gain outdoor unit thermistor readings to calculate roughly how the system is operating, in sort of a green/yellow/red indicator fashion. I found some information: [https://hvacbasics.com/non-invasive-testing/](https://hvacbasics.com/non-invasive-testing/)

For my model FTQ-PB (Outdoor RZQ-P) the only regularly reported information over p1p2 is very basic status is for the thermostat (that I can see so far) but there is a request type packet to obtain outdoor unit thermistor status. I don't know if additional temperature sensors might be necessary as well.

**Comment by Arnold-n** [comment]
Have you seen packets on P1/P2 to request the outdoor unit status? That would be interesting. Not sure if the auxiliary controller could also initiate these requests (for E-series systems this is not always possible, so P1P2Monitor simply inserts messages even if it is not intended by the protocol). Alternatively you could monitor the F1/F2 line between indoor and outdoor unit to see if any thermistor related messages are regularly communicated - but only if the F1/F2 lines have a DC voltage of at least 10V and at most 25V.

**Reply by gmcmicken** [reply]
Yes, using the hidden menu on the BRC controller, and selecting "outdoor unit status", I can see a request packet every second while the screen is open. One response packet with thermistor data is returned.

**Reply by gmcmicken** [reply]
```
Request:
00 00 A3
Payload:
99

Response
40 00 A3 
Payload:
80000080000080000040085F002A9880000000C8
```

**Reply by Arnold-n** [reply]
Ah, so given the 800000 pattern the 80* packets seen after boot-up may also just be forwarded packets from the outside unit. I guess the real content here is in 40 08 5F 00 2A 98, perhaps as 2x3 or 3x2 bytes? The last bytes 99 and C8 are CRC bytes and not part of the actual payload.

**Reply by gmcmicken** [reply]
There are blank thermistors slots on the screen too so I'm not sure what the order and offsets could be.
![16720036541068414296113381781277](https://user-images.githubusercontent.com/7595368/209482340-d1eadd4d-8a43-4535-9393-9f015602e651.jpg)

**Reply by Arnold-n** [reply]
I don't recognize a relation yet between raw data and thermistor values. Perhaps Fahrenheit or raw sensor value?

Just made an experimental branch experimental-FTQ-PB to cycle-steal the 40F030 response space to request thermistor status (transmit 0000A3 once ("C1") or every minute ("C2" - but C2 may not work on FTQ due to lack of clock)) - this could lead to bus collissions depending on how the BRC responds to this out-of-protocol behaviour when the FTQ will respond. Please monitor error status and P1P2/R/# when using this. Hope it compiles and works (cannot test it from where I am).

**Reply by gmcmicken** [reply]
Sorry to confuse you, these were different instances in time. The values don't match the screen, just the positions.

**Reply by gmcmicken** [reply]
I tried the experimental branch but unfortunately P1P2Monitor responded to every known 3x packet type with all FF the entire payload. It was a disaster and it partially scrambled the BRC's brain! I even lost the backlight on it. There are some values that flowed through to the heat pump itself in packet types 10, 11, and 20. I'm trying to compile a list now of what the values were and what they are now to see if I can restore them back but I don't have a way to write these (I don't think?)

```
P1P2/R/080 R T  0.156: 00F038013F6101200057821F005705150A38C4381A1FF036
P1P2/R/080 R T  0.030: 40F038FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFC9
P1P2/R/080 R T  0.155: 00F03A02000010078017800030808043
P1P2/R/080 R T  0.030: 40F03AFFFFFFFFFFFFFFFFFFFFFFFF9D
P1P2/R/080 R T  0.147: 00F0390000000000000000000000000095
P1P2/R/080 R T  0.030: 40F039FFFFFFFFFFFFFFFFFFFFFFFFFFEF
```

**Reply by Arnold-n** [reply]
Apologies that this happened, I think because P1P2Config.h still had E_SERIES defined. For E-series this all-FF reply to most of the 00F03x packets is expected behaviour in L1 mode. This definitely requires a safeguard to avoid happening again on F models as E_SERIES is defined by default! I will take this experimental branch offline and undefine E_SERIES in all config files so the code does not compile without making a decision for one of the system types.

Can you make a longer log of the current data communication and send me privately? We may have to write some code to have the interface act as BRC. Also, have you tried rebooting the system to see which values do not change back to o/insertriginal ones?

There is a - also very experimental - command to write individual 000010* / 000011*  messages using the W command in L1 mode. It cycle-steals one 00F030 reply slot to send one individual message. It has only been tried once on my system with other packet types.  This might work to restore the original values, but the BRC may overwrite the values if it doesn't accept the new heat pump replies. it requires the definition of ENABLE_INSERT_MESSAGE in P1P2Config.h, but again: very experimental, results uncertain.

**Reply by gmcmicken** [reply]
I sent you an email. No need to jump on this I was able to restore most, if not all, of it using the BRC and shutting off power to the heat pump. 

I did comment out E_SERIES before I compiled, or I thought I did.

But hey, interestingly I read about the BRC backlight (it's working again after the power cycle) that actually the bus doesn't have enough power for two controllers both with their backlights on. So the first controller to turn on the backlight sends a message to the other, and that controller is locked with the backlight off. So if we want to know how much power the bus has to offer... It is the amount of two BRC controllers and one backlight.

**Reply by Arnold-n** [reply]
I cannot see how an all-FF-reply would happen without E_SERIES defined; if so we would need to track that cause down. Would you still have the hex file or even better the *.lst and *.map files from that compilation?

Interesting about the back-light, would be nice to see which bits/bytes communicate LCD status. My Daikin thermostat consumes  (by heart) around 40mA with LCD off and 20mA more with LCD on; the P1P2-ESP-interface consumes 31mA at 15V without w5500, or 56mA with w5500, but with long/thin wires voltage goes down and current goes up.

**Reply by gmcmicken** [reply]
Okay so it was my fault, when I upgraded the Arduino IDE I had auto save on, but I turned it off at some point when I was working on something (I think the undo/redo was giving me trouble so I didn't want to save the file until I was ready)

Anyway I changed the config before compiling but didn't save it, and autosave was off.

**Reply by gmcmicken** [reply]
The new IDE works well with git though, so that's a major improvement!

**Reply by gmcmicken** [reply]
> There is a - also very experimental - command to write individual 000010* / 000011* messages using the W command in L1 mode. It cycle-steals one 00F030 reply slot to send one individual message. It has only been tried once on my system with other packet types. This might work to restore the original values, but the BRC may overwrite the values if it doesn't accept the new heat pump replies. it requires the definition of ENABLE_INSERT_MESSAGE in P1P2Config.h, but again: very experimental, results uncertain.

Can you give an example? If I issue "w 0000A3" will it send message 0000A399?

**Reply by Arnold-n** [reply]
Indeed. It will send that message exactly once (after a 00F030 message, in the time slot where P1P2Monitor is supposed to send 40F030*), and you will need to observe/log P1P2/R to see the 4000A3* reply (if any).

**Reply by gmcmicken** [reply]
Okay so I tried "w 0000A3" and it queued it up but didn't write it as I guess the 00F030 message is only seen when there is not a secondary controller already exchanging 00F038? So I used "L0" then "w 0000A3" and I got the message inserted, seemed to work fine.. Obviously I probably don't want to toggle L0/L1 all the time to put this in though so what is your suggestion?

**Comment by gmcmicken** [comment]
Okay I'm trying to add packet 4000A3 to ParameterConversion_F, this is the response packet:
`4000A380000080000080000040085F002A9880000000C8`
and I think (from the screen) it contains 6 thermistor readings (but mine reports only two).
```
4000A3 800000 800000 800000 40085F 002A98 800000 00C8
         1      2      3      4      5      6
```

Do you have a suggestion how to implement?

**Reply by gmcmicken** [reply]
Screenshot and packet at same point-in-time:
`4000A3 800000 800000 800000 000DAC 00145A 800000 0099`
![image](https://user-images.githubusercontent.com/7595368/218279796-ecf8390f-1858-4e6a-8479-30b9d9b48002.png)

**Reply by gmcmicken** [reply]
And one with negative values:
`4000A3 800000 800000 800000 40122D 002767 800000 00F2`
![image](https://user-images.githubusercontent.com/7595368/218282470-698f26c8-90a8-44fa-9e0c-772414b734cd.png)

**Reply by Arnold-n** [reply]
Thanks, helpful combi! I added the thermistors to the HA output; thermistors are not output if first byte is 0x80.
The v0.9.33 P1P2Monitor for F-series (just uploaded) now uses *any* 00F03x time slot for cycle-stealing, please check logs whether this does not give any bus collissions or other trouble.

**Reply by Arnold-n** [reply]
The last payload byte 0x00 could match the (outdoor) unit no, so if anyone has multiple outdoor units, I'd be interested in logs, otherwise I'll just keep the code simple for one outdoor unit only.

**Reply by gmcmicken** [reply]
Thanks! I tried to cycle the "unit number" to see if that byte changed but it won't let me cycle (I suppose it's smart enough to not allow if you don't have another unit connected).

I'll flash these and give it a try.

**Reply by gmcmicken** [reply]
I'm not seeing th4 or th5 come into home assistant, I removed all the config (even mqtt integration) and added it back. I'm expecting to see T1_Th4 and T1_Th5 in homeassistant/sensor config but it's not there.

Anyway I temporarily set up mqtt sensor on topic P1P2/P/#/T/1/Th4 and 5, as that is working.

**Reply by Arnold-n** [reply]
Forgot to update the history code, should work now.

**Reply by gmcmicken** [reply]
Thanks! It looks like the 'w' command is still waiting for 00F030 packet only (not 3x as you described), maybe just an error in #define ENABLE_INSERT_MESSAGE_3x
